Clinical-Genomics/microSALT
# Stability bundle

# Description
The features of this PR primarily concerns internals.

End-users:
- Responsible bioinformatician added to both logs and final report
- Json output reports now default to '' (empty) for non-standard values

Bioinformatics:
- Reports aren't removed from current folder, if a report is already present

Internal:
- The index checker for blast databases is now more thorough
- BLAST searches generalized to single function
- EXPAC set correctly renamed to ExPEC
- Optimized loci-length checker
- Fixed oversights in verified_fastq function, and database indexes when using parse_seqids
- CSV files are now readable regardless of language version
- Kmer coverage is used as a tertiary sorter if identity and span are otherwise identical
- Fixed an issue where some characters were incorrectly displayed when printing a report
- Input fastq are now verified to not end improperly


The reason for packaging these features is due to shifting focus in development, and packaging them avoids the code from becoming legacy/forgotten/useless.

## Primary function of PR

# Testing
_This is a description of the tests necessary to verify the stability of the PR._

- _`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-microsalt-stage.sh json_fix`_
- _`us`_
- _`source activate S_microSALT`_
- _`microSALT analyse project MIC3109`_
- _`microSALT analyse project ACC5551`_

_Verify that the results for projects MIC3109 & ACC5551 are consistent with the results attached to AMSystem doc 1490, Microbial_WGS.xlsx_

Additionally:
- Make sure responsible bioinformatician is written within the Typing & QC reports
- Use `microSALT utils report <PROJECT> --type json_dump` and verify that the non-standard values (like Control) are written as empty
- Use `microSALT utils report <PROJECT> --type default --output .` twice to verify that the report isn't removed

## Test results
_These are the results of the tests, and necessary conclusions, that prove the stability of the PR._

MIC3109 - QC: Coverage based metrics offset by less than 5%. Otherwise identical
ACC5551 - Typing results: Convergent
ACC5551 - Resistance results: Convergent

Running `microSALT utils report <PROJECT> --type json_dump` for project MIC3109 & ACC5551 produces json files where multiple values are ''; such as n50 and duplication_rate for project MIC3109.
Running `microSALT utils report <PROJECT> --type default --output .` twice still left the reports intact.


# Sign-offs
